### PR TITLE
(#280) Add test to exercise various existing / new print functions

### PR DESCRIPTION
**Pull request by gapisback**
This commit adds a new test case that exercises various functions to print different types of of pages, BTree innards, and RC-allocator innards.

- Add a function to print the super-block
- Modularize BTree printing function to sub-fns for diff page types
- Modularize Trunk printing function to sub-fns for pivots, bundles etc
- Improve layout of allocator details, showing extent-ID, addr and count
- Rename several print-fns to _sprint*(), taking an output stream handle.
- Minor corrections to some trunk print fns for proper alignment of output

All output from this test case will now be go to the /tmp file where tests redirect stdout.

### CODE REVIEW

**Comment by gapisback** [comment]
Hi, @ajhconway , @rtjohnso - This oldish change-set which ran into print-function naming conventions flak has been overhauled now. Here is what I did:

- Rebased v/s /main, picking up Alex's rework of `platform_log()` and related changes done under SHA 235fae2
- Reworked the set of new print functions that I had previously carved out of existing code to conform to the new naming / interface conventions introduced recently.

As an example, here are some new interfaces after this rework:

```
trunk_print_super_block(platform_log_handle *log_handle, trunk_handle *spl)
trunk_print_memtable(platform_log_handle *log_handle, trunk_handle *spl)
+trunk_print_branches_and_bundles(platform_log_handle *log_handle,
+                                 trunk_handle        *spl,
+                                 page_handle         *node)

+btree_print_leaf_node(platform_log_handle *log_handle,
+                      btree_config        *cfg,
+                      uint64               addr,
+                      btree_hdr           *hdr)
```
... and so on.

- No more will you see those `_sprint_` interfaces, using `stream` handles.
- I had to fuss around with the output widths as due to the addition of `kv bytes` in PIVOTS output block, the width becomes slightly longer. Had to fuss around this in a few diff places.

The best way to re-review this would be to take a look at the [diffs in this commit](https://github.com/vmware/splinterdb/pull/281/files#diff-3270c318319a7ea42847124970afae8d01428ef1e4efc80a40809c1ca24ebe3b). That is the one where the reconciliation of these new print interfaces w/ Alex's rework can be found.

There are still a couple of minor niggles, which I will comb through and annotate in the diffs. They are minor; if we don't resolve them in this round, it's ok. I'll poke you off-line when we meet.

------

**NOTE**: As an alert, ... I have added this page to our docs [Diagnostics & Troubleshooting](https://github.com/vmware/splinterdb/blob/agurajada/280-test-print-diags/docs/diagnostics.md), and see that the href-links are a broken now. 

There is a bit of docs cleanup to be done due to this code rework. Will fix them all before merging. 

I want to use this page as a way to communicate to users "some" details of our internal layouts.

Thanks. ... If you can take a quick 2nd look, I am ready to integrate this pending item.

File: src/trunk.c [hunk 1]
```diff
@@ -1778,6 +1821,11 @@ trunk_inc_num_pivot_keys(trunk_handle *spl, page_handle *node)
    trunk_hdr *hdr = (trunk_hdr *)node->data;
    debug_assert(hdr->num_pivot_keys >= 2);
    hdr->num_pivot_keys++;
+   /* RESOLVE: Do we have an off-by-1 issue here?
+    * In trunk_add_pivot(), we check
+    * if (trunk_num_pivot_keys() > cfg.max_pivot_keys) and then
+    * raise an error. Should we not raise an error here, too?
+    */
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/trunk.c:1918]: @rtjohnso - I took a 2nd look.  You are right; I misunderstood. The 2nd place where I flagged as an error is being raised is here; 

```
1456 trunk_add_pivot(trunk_handle *spl,
[...]
1465    if (trunk_num_pivot_keys(spl, parent) >= spl->cfg.max_pivot_keys) {
1466       // No room to add a pivot
1467       debug_assert(trunk_num_pivot_keys(spl, parent)
1468                    == spl->cfg.max_pivot_keys);
1469       return STATUS_LIMIT_EXCEEDED;
1470    }
```

So, yes, they are logically equivalent checks.

File: src/trunk.c [hunk 2]
```diff
@@ -7905,68 +7968,109 @@ trunk_print_locked_node(platform_log_handle *log_handle,
...
             // clang-format off
-            platform_log(log_handle, "|     |        -- pivot %2u --       |              |              |              |                |\n",
-                                pivot_no);
+            platform_log(log_handle, "|     |        -- pivot %2u --       |              |              |              |                 |\n",
+                         pivot_no);
             // clang-format on
          }
       }
+
+      // Iterate through all the bundles ...
```
Comment (ajhconway) [inline src/trunk.c:8053]: This comment is confusing, since this is a nested iteration, and all we're really doing is searching for bundles that start at this branch.

File: src/trunk.c [hunk 3]
```diff
@@ -7542,15 +7654,23 @@ trunk_print_locked_node(trunk_handle          *spl,
...
+// clang-format on
+
+/*
+ * trunk_sprint_branches_sb_filters() -- Flesh this out.
+ */
+static void
+trunk_sprint_branches_sb_filters(platform_stream_handle stream,
+                                 trunk_handle          *spl,
+                                 page_handle           *node)
+{}
```
[the remark this answers is not in the document]
Comment (ajhconway) [reply src/trunk.c:8110]: I still don't know what you want do to here, but I don't see anything sensible to print, so I would remove this function.

File: src/trunk.c [hunk 4]
```diff
@@ -8025,14 +8145,21 @@ trunk_print_subtree(trunk_handle        *spl,
...
    }
    cache_unget(spl->cc, node);
 }
 
+/*
+ * trunk_print_memtable() --
+ *
+ * Print the currently active Memtable, and the other Memtables being processed.
+ * Memtable printing will drill-down to BTree printing which will keep
+ * recursing. So ... you'll end up with TONS of output; all very useful.
```
Comment (ajhconway) [inline src/trunk.c:8159]: Kill this last sentence.

I agree this isn't a useful function in general and we should consider just removing it.

File: src/trunk.c [hunk 5]
```diff
@@ -8046,20 +8173,58 @@ trunk_print_memtable(trunk_handle *spl, platform_log_handle *log_handle)
...
+      memtable_print(log_handle, spl->cc, mt);
    }
    platform_log(log_handle, "\n");
 }
 
+/*
+ * trunk_print()
+ *
+ * Driver routine to print a SplinterDB trunk, and all its sub-pages.
+ * Caller can supply non-NULL user_stream handle to redirect output.
```
Comment (ajhconway) [inline src/trunk.c:8191]: This sentence doesn't make sense. Caller must supply a non-NULL handle. Moreover, what is a "user_stream handle"?

File: src/trunk.h [hunk 6]
```diff
@@ -396,6 +396,13 @@ trunk_reset_stats(trunk_handle *spl);
 
 void
 trunk_print(platform_log_handle *log_handle, trunk_handle *spl);
+
+void
+trunk_print_super_block(platform_log_handle *log_handle, trunk_handle *spl);
+
+void
+trunk_sprint(platform_stream_handle user_stream, trunk_handle *spl);
+
```
Comment (ajhconway) [inline src/trunk.h:405]: Again wasn't this going to change?

File: tests/unit/splinter_test.c [hunk 7]
```diff
@@ -428,10 +429,11 @@ CTEST2(splinter, test_lookups)
 
    trunk_shadow shadow;
    trunk_shadow_init(&shadow);
+   uint64 num_inserts = 0;
 
    // FALSE : No need to do verification-after-inserts, as that functionality
    // has been tested earlier in test_inserts() case.
-   uint64 num_inserts = splinter_do_inserts(data, spl, FALSE, &shadow);
+   num_inserts = splinter_do_inserts(data, spl, FALSE, &shadow);
```
Comment (ajhconway) [inline tests/unit/splinter_test.c:436]: Why this change? I generally strongly prefer defining variables where they are used.

File: tests/unit/splinter_test.c [hunk 8]
```diff
@@ -623,6 +625,58 @@ CTEST2(splinter, test_lookups)
...
+                        "Generated by %s:%d:%s ****\n",
+                        __FILE__,
+                        __LINE__,
+                        __FUNCTION__);
+
+   trunk_print_super_block(Platform_default_log_handle, spl);
+
+   trunk_print_space_use(Platform_default_log_handle, spl);
+
+   platform_default_log("\n** trunk_sprint() **\n");
```
Comment (ajhconway) [inline tests/unit/splinter_test.c:665]: Again here...

File: tests/unit/splinter_test.c [hunk 9]
```diff
@@ -509,6 +509,57 @@ CTEST2(splinter, test_lookups)
    trunk_destroy(spl);
 }
 
+/*
+ * Simple test cases to exercise print / diagnostic functions provided
+ * by various sub-systems.
+ */
+CTEST2(splinter, test_splinter_print_diags)
```
[the remark this answers is not in the document]
Comment (ajhconway) [reply tests/unit/splinter_test.c:637]: What does this test do? It just prints the stuff right? But it doesn't succeed or fail based on the output, right? So it's a visual test?

File: tests/unit/splinter_test.c [hunk 10]
```diff
@@ -668,6 +722,7 @@ splinter_do_inserts(void         *datap,
    // Debug hook: Override this to smaller value for faster test execution,
    // while doing test-dev / debugging. Default is some big value, like
    // 12988800; num_inserts = (100 * 100 * 100);
+   // num_inserts = (100 * 100 * 100);
```
Comment (ajhconway) [inline tests/unit/splinter_test.c:725]: Rebase artifact?

**Review by ajhconway — CHANGES_REQUESTED** [review]
Some things have been resolved, others haven't.

File: docs/diagnostics.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
+# Diagnostics and Troubleshooting
+This document gives brief references to diagnostics and troubleshooting
+interfaces available in the library.
+
+A small collection of print routines is available which can be used to examine key data strucures, such as the Splinter trunk pages, BTree pages, space-allocation maps and so on.
+
+- [trunk_print_super_block](../src/trunk.c#:~:text=trunk%5Fprint%5Fsuper%5Fblock%28)
+  Print contents of Splinter super block
```
[the remark this answers is not in the document]
Comment (gapisback) [reply docs/diagnostics.md:8]: Done.

File: docs/diagnostics.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+- [trunk_print](../src/trunk.c#:~:text=trunk%5Fprint%28)
+  Print contents of a SplinterDB trunk and all its sub-pages 
+
+- [trunk_print_node](../src/trunk.c#:~:text=trunk%5Fprint%5Fnode%28)
+  Print contents of a Trunk node (page)
+
+- [trunk_print_subtree](../src/trunk.c#:~:text=trunk%5Fprint%5Fsubtree%28platform%5Flog%5Fhandle)
+  Print contents of all sub-trees under a Trunk node (page)
```
[the remark this answers is not in the document]
Comment (gapisback) [reply docs/diagnostics.md:17]: Well ... I wasn't exactly trying to list out an extern / internal ref-man for print functions.

I was simply trying to "list out" all existing print functions, that devs who wish to get into internals can navigate to from this docs page. 

It's sort-of like a **quick-start cheat-sheet**, without having the users troll thru the code to figure out how to print internal structures.

I feel. ... trying to put more structure on this at this time, may end up creating more work for us.
Let's start with this ... and if things evolve, or depending on any reactions from users, we can re-lay this out.

File: docs/diagnostics.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+- [trunk_print_space_use](../src/trunk.c#:~:text=trunk%5Fprint%5Fspace%5Fuse%28)
+  Print space usage by level across all trunk nodes
+
+- [trunk_print_memtable](../src/trunk.c#:~:text=trunk%5Fprint%5Fmemtable%28platform%5Flog%5Fhandle)
+  Print contents of all Memtables in Splinter. Drills-down to printing BTrees under Memtables
+
+- [btree_print_node](../src/btree.c#:~:text=btree%5Fprint%5Fnode%28platform%5Flog%5Fhandle),
+  [btree_print_locked_node](../src/btree.c#:~:text=btree%5Fprint%5Flocked%5Fnode%28platform%5Flog%5Fhandle)
+  Print BTree nodes, and recurse down to sub-tree nodes.
```
[the remark this answers is not in the document]
Comment (gapisback) [reply docs/diagnostics.md:27]: OK, point noted. Will address this before I check this in.

(To do docs updates, I have to upload the md-files, and cross-check href-links. Will make that cleanup in a final pass.)

Also, again, ... I will include `btree_tree_print`, but I am not trying to stick to extern print fns alone. I am seeing this as a rich library of all useful print fns' entry points.

File: docs/diagnostics.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+- [rc_allocator_print_debug](../src/rc_allocator.c#:~:text=rc_allocator%5Fprint%5Fdebug%28%29),
+  Print the space allocation maps and lists of extents / pages allocated.
+
+- [srq_print](../src/srq.h#::~:text=srq_print%28%29)
+  Print contents of the space reclamation queue
```
[the remark this answers is not in the document]
Comment (gapisback) [reply docs/diagnostics.md:33]: Deleted this bullet.

File: src/btree.c [hunk 15]
```diff
@@ -64,6 +64,21 @@ slice positive_infinity = {0, &positive_infinity_buffer};
...
+                       btree_hdr           *hdr);
+
+static void
+btree_print_leaf_node(platform_log_handle *log_handle,
+                      btree_config        *cfg,
+                      uint64               addr,
+                      btree_hdr           *hdr);
+
+static void
+btree_print_offset_table(platform_log_handle *log_handle, btree_hdr *hdr);
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/btree.c:81]: Wasn't sure if you really wanted these moved, or not. 

This is really a C-ism ... The prototypes help to define the function after its callee, which somewhat helps with readability -- you read the functions in the order of their flow of execution.

I had a running argument on this with @rtjohnso also ... who also seems to prefer your style. (We can talk off-line.)

Much as I don't quite prefer that style (as it impairs readability) -- I did rearrange them to not require the prototypes.

File: src/btree_private.h [hunk 16]
```diff
@@ -315,6 +315,11 @@ btree_get_tuple_message_type(const btree_config *cfg,
 }
 
 
+/*
+ * Return a ptr to the k'th index_entry on the BTree header page.
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/btree_private.h:319]: Done; corrected.

File: src/btree_private.h [hunk 17]
```diff
@@ -315,6 +315,11 @@ btree_get_tuple_message_type(const btree_config *cfg,
...
+ * The byte offset of this entry from the start of the header page is given
+ * by the offsets[k]'th value. Add diagnostic checks to validate addresses.
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/btree_private.h:321]: Good idea about the comment. Moved it to the top of this file, where the struct is declared, and reworded.

The comment is simply trying to say that this function does some extra validation. Reworded to:

`Validates addresses, in debug mode.`

File: src/rc_allocator.c [hunk 18]
```diff
@@ -161,13 +161,13 @@ rc_allocator_print_stats_virtual(allocator *a)
 }
 
 void
-rc_allocator_debug_print(rc_allocator *al);
+rc_allocator_print_debug(rc_allocator *al);
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/rc_allocator.c:164]: I settled for `rc_allocator_print_allocated()`, as we are any printing "all the allocated" extents. So `_all_` seemed superfluous. (And we don't quite have a way to print only `_some_` allocated extents.)

File: src/rc_allocator.c [hunk 19]
```diff
@@ -161,13 +161,13 @@ rc_allocator_print_stats_virtual(allocator *a)
...
 
 void
 rc_allocator_debug_print_virtual(allocator *a)
 {
    rc_allocator *al = (rc_allocator *)a;
-   rc_allocator_debug_print(al);
+   rc_allocator_print_debug(al);
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/rc_allocator.c:170]: Done. Changed to this:

`167 rc_allocator_print_allocated_virtual(allocator *a)`

File: src/rc_allocator.h [hunk 20]
```diff
@@ -115,4 +115,10 @@ rc_allocator_mount(rc_allocator        *al,
 void
 rc_allocator_dismount(rc_allocator *al);
 
-#endif
+void
+rc_allocator_sprint_stats(platform_stream_handle stream, rc_allocator *al);
+
+void
+rc_allocator_sprint_debug(platform_stream_handle stream, rc_allocator *al);
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/rc_allocator.h:122]: Sorry, my mistake!

Those two functions have now been withdrawn, as the allocated extents printing is being piped thru the virtual function. 

These are straggling prototypes for non-existent functions. Deleted.

File: src/trunk.c [hunk 21]
```diff
@@ -89,6 +89,13 @@ static const int64 latency_histo_buckets[LATENCYHISTO_SIZE] = {
  */
 #define TRUNK_SINGLE_LEAF_THRESHOLD_PCT (75)
 
+/*
+ * During Splinter configuration, the fanout parameter is provided by the user.
+ * We slightly over-provision the max # of pivot keys on a trunk node by this
+ * value.
+ */
+#define TRUNK_EXTRA_PIVOT_KEYS (6)
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/trunk.c:97]: Done; dropped this as-is to the comment.

File: src/trunk.c [hunk 22]
```diff
@@ -306,21 +313,21 @@ static const int64 latency_histo_buckets[LATENCYHISTO_SIZE] = {
 
 /*
  *-----------------------------------------------------------------------------
- * Trunk Nodes: splinter_trunk_hdr{}: Disk-resident structure
+ * Trunk Nodes: Splinter trunk_hdr{}: Disk-resident structure
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/trunk.c:345]: Reverted. ( ... may be... `SplinterDB trunk_hdr{} ...` would read better ... ?

File: src/trunk.c [hunk 23]
```diff
@@ -406,7 +420,7 @@ trunk_log_node_if_enabled(platform_stream_handle *stream,
  *          soft limit. The hard limit (hard_max_branches_per_node) is the
  *          number of branches (whole and fractional) for which there is
  *          physical room in the node, and as a result cannot be exceeded. An
- *          attempt to flush into a node which is at the hard limit will fail.
+ *          attempt to flush _into_ a node which is at the hard limit will fail.
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/trunk.c:423]: `_into_` is stylistic ascii-speak for _**emphasis**_.

As in, "flush into a node" vs/ "flush **into** a node"

File: src/trunk.c [hunk 24]
```diff
@@ -1186,18 +1209,27 @@ trunk_inc_pivot_generation(trunk_handle *spl, page_handle *node)
...
+/* struct trunk_pivot_data{} itself is the message. Return its size */
 uint64
 trunk_pivot_message_size()
 {
    return sizeof(trunk_pivot_data);
 }
 
+/*
+ * Return the start of the pivot_no'th pivot_data, which begins just past
+ * the key. (Key size is effectively a statically configured limit.)
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/trunk.c:1231]: Well, ... not so sure. 

Usually the term "runtime" means this limit can change across reboots. Say, like the cache-size.

But once you have configured the `data_cfg` with a key-size, that gets blazed into the SplinterDB config structs, so you cannot change it. Hence, i chose 'statically'.

Once when we do support variable key-sizes, I'm expecting `trunk_key_size()` itself to change or go away.
At that time, this comment may have to be reworked.

I know that sounds like leaving extra clean up behind, but the intent of the comment is to tighten and clarify exactly what this key-size is all about **today**.

File: src/trunk.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (gapisback) [reply src/trunk.c:1918]: Yes, you are right. My mistake; sorry!

Deleted.

File: src/trunk.c (hunk 2 above)
Comment (gapisback) [reply src/trunk.c:8053]: Deleted existing comment, and reworded as you described.

File: src/trunk.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (gapisback) [reply src/trunk.c:8110]: Done. (I was hoping to figure something out and write it here,  ... but ran out of time. Will revisit another day.)

File: src/trunk.c (hunk 4 above)
Comment (gapisback) [reply src/trunk.c:8159]: Let's retain the function, I like to see all the internals printed, at least as an educational tool.  I think it's good to have such tools that dump everything out.

I deleted the cautionary comment.

File: src/trunk.c (hunk 5 above)
Comment (gapisback) [reply src/trunk.c:8191]: Deleted. Was a left-over comment from prior version of this print interface. (Thanks for catching it!)

File: src/trunk.h (hunk 6 above)
Comment (gapisback) [reply src/trunk.h:405]: Straggling reference from previous version. Deleted. [ Sorry ! ]

File: tests/unit/splinter_test.c (hunk 7 above)
Comment (gapisback) [reply tests/unit/splinter_test.c:436]: Don't recall why I edited this. Grubby fingers. Reverted.

File: tests/unit/splinter_test.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (gapisback) [reply tests/unit/splinter_test.c:637]: Yes, in a way it's a visual test.

But the expectation is -not- that the dev will visually check this for correctness.

All I'm doing is **exercising** the print methods, in their full glory, with a biggish db.

This way, if ever the print methods change, or something causes them to seg-fault we will find it in CI-runs.

Including this test case, thereby, provides a small safety net against print code regressing unknowingly.

File: tests/unit/splinter_test.c (hunk 8 above)
Comment (gapisback) [reply tests/unit/splinter_test.c:665]: fixed. [ really, sorry ! I did think I had grepped for all instances of `_sprint_` before publishing this reworked review. ]

File: tests/unit/splinter_test.c (hunk 10 above)
Comment (gapisback) [reply tests/unit/splinter_test.c:725]: Not exactly.

In another review, you had frowned upon the use of variables named `shadowp` or `shadowpp`.

The code had been fixed, previously, but the prolog comment was not. I am just fixing the name `shadowpp`.

(Btw ..., as an aside, it does help with understanding to see `shadowpp`, especially if it's an output parameter, where caller is passing-in addr-of-ptr-to-struct to be filled-out.)

**Comment by gapisback** [comment]
@ajhconway - Thanks for a careful tooth-comb re-review. I fixed the straggling pieces of prototypes and various other cleanup you suggested.

The docs file .... I will re-check the href-links one final time before I merge; leave them for now.

Let me know when you've had a re-re-look, and pass this. I'd rather wait till this is finalized, as this change set has been a bit controversial. And want to make sure that it finally conforms all the required naming conventions.

**Comment by gapisback** [comment]
About the test itself: It is there to ensure that the print methods are always working reliably, at least for some big'gish input set.

Otherwise, if they seg-fault when you really need them, that will be very awkward. It's **mostly a regression test** against less-used print methods; as a **defense check against bit-rot.**

File: tests/unit/splinter_test.c (hunk 10 above)
Comment (gapisback) [reply tests/unit/splinter_test.c:724]: I missed the line you remarked on. Deleted L725. I had added it for my own dev-testing.

File: docs/diagnostics.md [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+- [trunk_sprint_subtree](../src/trunk.c#:~:text=trunk\_sprint\_subtree\(platform\_stream\_handle)
+  Print contents of all sub-trees under a Trunk node (page)
+
+- [trunk_print_space_use](../src/trunk.c#:~:text=trunk\_print\_space\_use\(trunk\_handle)
+  Print space usage by level across all trunk nodes
+
+- [trunk_sprint_memtable](../src/trunk.c#:~:text=trunk\_sprint\_memtable\(platform\_stream\_handle)
+  Print contents of all Memtables in Splinter. Drills-down to printing BTrees under Memtables
+
+- [btree_print_node](../src/btree.c#:~:text=btree\_print\_node\(cache),
```
Comment (ajhconway) [inline docs/diagnostics.md:24]: Should probably also mention `btree_print_tree` and maybe `btree_print_lookup`.

File: docs/diagnostics.md [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+  [btree_print_locked_node](../src/btree.c#:~:text=btree\_print\_locked\_node\(btree\_config)
+  Print BTree nodes, and recurse down to sub-tree nodes.
+
+- [rc_allocator_print_debug](../src/rc_allocator.c#:~:text=rc_allocator\_print\_debug\(rc_allocator),
+  [rc_allocator_sprint_debug](../src/rc_allocator.c#:~:text=rc_allocator\_sprint\_debug\(platform\_stream\_handle)
```
Comment (ajhconway) [inline docs/diagnostics.md:29]: Here's another `sprint`

File: src/btree.c [hunk 27]
```diff
@@ -3157,6 +3156,109 @@ btree_count_in_range_by_iterator(cache        *cc,
...
+
+   uint64 nentries = btree_num_entries(hdr);
+   char   fmtstr[30];
+   snprintf(fmtstr,
+            sizeof(fmtstr),
+            "[%%%s] %%-8u",
+            ((nentries < 10)     ? "d"
+             : (nentries < 100)  ? "2d"
+             : (nentries < 1000) ? "3d"
+                                 : "4d"));
```
Comment (ajhconway) [inline src/btree.c:3174]: This feels unnecessary and tedious. Why not just use `3d` (1000s of entries feels unlikely to happen soon)?

File: src/btree.c [hunk 28]
```diff
@@ -3157,6 +3156,109 @@ btree_count_in_range_by_iterator(cache        *cc,
...
+
+static void
+btree_print_index_node(platform_log_handle *log_handle,
+                       btree_config        *cfg,
+                       uint64               addr,
+                       btree_hdr           *hdr)
+{
+   data_config *dcfg = cfg->data_cfg;
+   platform_log(log_handle, "**  INDEX NODE \n");
+   platform_log(log_handle, "**  hdrptr: %p\n", hdr);
```
Comment (ajhconway) [inline src/btree.c:3194]: I prefer printing `Header pointer:`

File: src/trunk.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (ajhconway) [reply src/trunk.c:425]: What's the point of the emphasis?

File: src/trunk.c [hunk 29]
```diff
@@ -1186,18 +1210,27 @@ trunk_inc_pivot_generation(trunk_handle *spl, page_handle *node)
...
+ * A pivot consists of the pivot key (of size cfg.key_size) followed by
+ * a struct trunk_pivot_data. Return the total size of a pivot.
+ */
 uint64
 trunk_pivot_size(trunk_handle *spl)
 {
    return trunk_key_size(spl) + sizeof(trunk_pivot_data);
 }
 
+/* struct trunk_pivot_data{} itself is the message. Return its size */
```
Comment (ajhconway) [inline src/trunk.c:1223]: This naming is confusing out of context and the comment doesn't really help. Maybe @rtjohnso could suggest something.

File: src/trunk.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (ajhconway) [reply src/trunk.c:1232]: That's fine, except that I'm taking issue with the word `static`, which implies compile-time. We're actually going out of our way to have this be determined at runtime, even though you are right that once set it is static across reboots. I still want the comment to clarify.

File: tests/unit/splinter_test.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (ajhconway) [reply tests/unit/splinter_test.c:636]: Ok, that sounds fine to me.

File: tests/unit/splinter_test.c (hunk 10 above)
Comment (ajhconway) [reply tests/unit/splinter_test.c:724]: I was referring to the some expression on L723.

**Review by ajhconway — APPROVED** [review]
There are some issues with `sprint` and `platform_stream_handle` in the doc file still, and there are some minor changes elsewhere I would like still. But after that I think this is ready.

**Review by rtjohnso — APPROVED** [review]

File: src/btree.c (hunk 27 above)
Comment (gapisback) [reply src/btree.c:3174]: Based on our off-line slack chat, changed this inline code to a macro, `PRInd()` defined in `util.h`.
Macro evaluates to one of the "d", "2d", etc. print format specifiers.

File: src/btree.c (hunk 28 above)
Comment (gapisback) [reply src/btree.c:3194]: Old code (before this work) was: `3180       platform_log(log_handle, "**  ptr: %p\n", hdr);`
Now Changed to: `3188    platform_log(log_handle, "**  Header ptr: %p\n", hdr);`

File: src/btree.c (hunk 27 above)
Comment (ajhconway) [reply src/btree.c:3174]: I like the macro, but can we call it something more in line with our naming conventions, like `DECIMAL_STRING_WIDTH`?

File: docs/diagnostics.md (hunk 25 above)
Comment (gapisback) [reply docs/diagnostics.md:24]: Done. Added.

File: docs/diagnostics.md [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+- [trunk_print](../src/trunk.c#:~:text=trunk\_print\_node\(trunk\_handle)
+  Print contents of a Trunk node (page)
+
+- [trunk_sprint_subtree](../src/trunk.c#:~:text=trunk\_sprint\_subtree\(platform\_stream\_handle)
+  Print contents of all sub-trees under a Trunk node (page)
+
+- [trunk_print_space_use](../src/trunk.c#:~:text=trunk\_print\_space\_use\(trunk\_handle)
+  Print space usage by level across all trunk nodes
+
+- [trunk_sprint_memtable](../src/trunk.c#:~:text=trunk\_sprint\_memtable\(platform\_stream\_handle)
```
[the remark this answers is not in the document]
Comment (gapisback) [reply docs/diagnostics.md:21]: cleaned up.

File: docs/diagnostics.md (hunk 26 above)
Comment (gapisback) [reply docs/diagnostics.md:29]: fixed.

**Comment by gapisback** [comment]
@ajhconway - I made all the corrections you pointed out. 

- Took out the comment about 'static' config for key-size, as expanding that would involve more re-work & re-review. And all this is anyway going to come out.

Am going in for the merge ... once CI passes.
